lib/build.ts

```diff
@@ -187,3 +194,169 @@ export async function runBuildTask(
+//
+//      NOTE!  Even when a build succeeds, if Docker is making assumptions then it is possible that
+//      the image will cause an execution error on the target platform if the architecture is not
+//      correct.
+//
+//   So what do we do?
+//
+//      + In the case of all images having a V2 manifest, we can simply pass the platform flag.
+//
+//      + In the case of having some V1 manifests ...
+//          - If we have balenalib images, we know that they have been updated.  Throw an error
+//            and tell the user to update
+//          - If we do not have balenalib images, don't pass the platform flag.  Let Docker
+//            close its eyes and hope for the best.  This opens the possibility for an exec
+//            error, but also allows users to continue to use v1 images if they want to / need to.
+//
+async function checkAllowDockerPlatformHandling(
+	task: BuildTask,
+	docker: Dockerode,
+): Promise<boolean> {
+	let imageReferences: string[];
+
+	const checkHasV1Manifest = async (repository: string) => {
+		try {
+			const manifest = await getManifest(docker.modem, repository);
+			return manifest.Descriptor.mediaType === MEDIATYPE_MANIFEST_V1;
+		} catch {
+			// eat the exception... yummy!
+		}
+
+		// at this time, treat any errors as true (v1 manifest)
+		return true;
+	};
+
+	if (task.imageName) {
+		imageReferences = [task.imageName + !!task.tag ? `:${task.tag}` : ''];
+	} else {
+		if (!task.dockerfile) {
+			// Sanity check
+			console.warn('Build task does not have an associated Dockerfile');
+			return false;
+		}
+		const { DockerfileParser } = await import('dockerfile-ast');
+		const parsedDockerfile = DockerfileParser.parse(task.dockerfile);
+		const dockerInstructions = parsedDockerfile.getInstructions();
+		const fromInstructions = dockerInstructions.filter(
+			(inst) => inst.getKeyword() === 'FROM',
+		);
+		if (fromInstructions.length === 0) {
+			// Sanity check
+			console.warn(
+				'Build task associated Dockerfile does no reference any images',
+			);
+			return false;
+		}
+		imageReferences = [];
+		fromInstructions.forEach((inst) => {
+			const fromLineArguments = inst.getArguments();
+			if (fromLineArguments.length === 0) {
+				// bail out, no image specified?
+				return;
+			}
+
+			imageReferences.push(fromLineArguments[0].getValue());
+		});
+	}
+
+	const outdatedBalenalibImages: string[] = [];
+	const v1Images: string[] = [];
+	const v2Images: string[] = [];
+
+	await Promise.all(
+		imageReferences.map(async (r) => {
+			const hasV1Manifest = await checkHasV1Manifest(r);
+			if (r.indexOf('balenalib/') >= 0) {
+				if (hasV1Manifest) {
+					outdatedBalenalibImages.push(r);
+				}
+			} else if (hasV1Manifest) {
+				v1Images.push(r);
+			} else {
+				v2Images.push(r);
+			}
+		}),
+	);
+
+	if (outdatedBalenalibImages.length > 0) {
+		throw new BuildProcessError(
+			'Error: The following balenalib images do not support platform targeting. Please update them:\n-' +
```

Comment:
I think a warning here is more realistic than forcing folks to update their pinned versions. We don't want to discourage version pinning in any way. What if the latest images break something for them?

Reply:
The error was something suggested by an earlier reviewer.  But I think I agree that it's a bit draconian.

Reply:
Given that we are doing this, we probably don't need to look for `balenalib` images at all.  V1 images will all be treated to the same warning.